src/concrete/ml/sklearn/base.py

```diff
@@ -1690,3 +1697,358 @@ def predict_proba(self, X: Data, fhe: Union[FheMode, str] = FheMode.DISABLE) ->
+
+                Args:
+                    x (numpy.ndarray): The encrypted input array
+                    indices (numpy.ndarray): The desired indexes
+
+                Returns:
+                    numpy.ndarray: The selected encrypted indexes.
+                """
+                arr = []
+                for i in indices:
+                    arr.append(x[i])
+                enc_arr = fhe_array(arr)
+                return enc_arr
+
+            def scatter1d(x, v, indices):
+                """Rearrange elements of `x` with values from `v` at the specified `indices`.
+
+                Args:
+                    x (numpy.ndarray): The encrypted input array in which items will be updated
+                    v (numpy.ndarray): The array containing values to be inserted into `x`
+                        at the specified `indices`.
+                    indices (numpy.ndarray): The indices indicating where to insert the elements
+                        from `v` into `x`.
+
+                Returns:
+                    numpy.ndarray: The updated encrypted `x`
+                """
+                for idx, i in enumerate(indices):
+                    x[i] = v[idx]
+                return x
+
+            def mul_tlu(a, b):
+                """Matrix multiplication.
+
+                Args:
+                    a (numpy.ndarray): An encrypted array
+                    b (numpy.ndarray): An encrypted array
+
+                Returns:
+                    numpy.ndarray: The result of a * b
+                """
+                return a * b
+
+            comparisons = numpy.zeros(x.shape)
+            idx = numpy.arange(x.size) + fhe_zeros(x.shape)
+
+            n, k = x.size, self.n_neighbors
+            ln2n = int(numpy.ceil(numpy.log2(n)))
+
+            for t in range(ln2n - 1, -1, -1):
+                p = 2**t
+                r = 0
+                d = p
+
+                for bq in range(ln2n - 1, t - 1, -1):
+                    q = 2**bq
+                    range_i = numpy.array(
+                        [i for i in range(0, n - d) if i & p == r and comparisons[i] < k]
+                    )
+                    if len(range_i) == 0:
+                        continue
+
+                    a = gather1d(x, range_i)  # x[range_i]
+                    a_i = gather1d(idx, range_i)  # idx[range_i]
+                    b = gather1d(x, range_i + d)  # x[range_i + d]
+                    b_i = gather1d(idx, range_i + d)  # idx[range_i + d]
+
+                    diff = a - b
+                    sign = diff < 0
+
+                    max_x = a + numpy.maximum(0, b - a)
+                    x = scatter1d(x, a + b - max_x, range_i)  # x[range_i] = a + b - max_x
+                    x = scatter1d(x, max_x, range_i + d)  # x[range_i + d] = max_x
+
+                    max_idx = a_i + mul_tlu((b_i - a_i), sign)
+
+                    # idx[range_i] = a_i + b_i - max_idx
```

Comment:
is this comment supposed to be removed or is it supposed to be next to the following line ?

Reply:
I think the comment explains the line by giving an alternative explanation. It could be re-written as :+1: 

`# The following is equivalent to:  idx[range_i] = a_i + b_i - max_idx`